fastn-core/src/migrations.rs

```diff
@@ -0,0 +1,224 @@
+    content: String,
+}
+
+async fn get_migrations_from_ftd(
+    req_config: &mut fastn_core::RequestConfig,
+) -> Vec<MigrationDataFTD> {
+    let migration_ftd = req_config
+        .config
+        .package
+        .fs_fetch_by_file_name("migration.ftd", None, &req_config.config.ds)
+        .await?;
+    let migration_ftd_str = String::from_utf8(migration_ftd).unwrap();
+
+    let main_ftd_doc = fastn_core::doc::interpret_helper(
+        "migration",
+        migration_ftd_str.as_str(),
+        req_config,
+        "/",
+        false,
+        0,
+    )
+    .await?;
+    let mut migration_data = vec![];
+    let mut bag = main_ftd_doc.data.clone();
+    bag.extend(ftd::interpreter::default::default_migration_bag());
+
+    let doc = ftd::interpreter::TDoc::new(&main_ftd_doc.name, &main_ftd_doc.aliases, &bag);
+
+    for (idx, component) in main_ftd_doc.tree.iter().enumerate() {
+        if is_fastn_migration(component) {
+            let property_values =
+                component.get_interpreter_property_value_of_all_arguments(&doc)?;
+            let query = property_values
+                .get("query")
+                .unwrap()
+                .clone()
+                .resolve(&doc, component.line_number)
+                .unwrap()
+                .to_json_string(&doc, false)?
+                .unwrap();
+            let migration_name = fastn_core::utils::generate_hash(&query);
+            let migration_number = (idx + 1) as i64;
+            migration_data.push(MigrationDataFTD {
+                number: migration_number,
+                name: migration_name,
+                content: query,
+            });
+        } else {
+            unimplemented!(); // todo: throw error
+        }
+    }
+
+    migration_data
+}
+
+fn is_fastn_migration(component: &ftd::interpreter::Component) -> bool {
+    component.name.eq("fastn#migration") || component.name.eq("fastn.migration")
+}
+
+struct MigrationDataSQL {
+    number: i64,
+    name: String,
+}
+
+async fn find_migrations_to_apply(
+    migration_ftd: &[MigrationDataFTD],
+    migration_sql: &[MigrationDataSQL],
+) -> Vec<MigrationDataFTD> {
+    if migration_sql.len() > migration_ftd.len() {
+        unreachable!("Can't delete the applied migration") // todo: throw error
+    }
+
+    let migration_sql: std::collections::HashMap<i64, MigrationDataSQL> = migration_sql
+        .into_iter()
+        .map(|val| (*val.number, val.clone()))
+        .collect();
+    let mut migrations_to_apply = vec![];
+    for m_ftd in migration_ftd {
+        match migration_sql.get(*m_ftd.number) {
+            Some(m_sql) => {
+                if m_sql.name.ne(&m_ftd.name) {
+                    unreachable!("Cannot change the content of migration") // todo: throw error
+                }
+            }
+            None => {
+                migrations_to_apply.push(m_ftd.clone());
+            }
+        }
+    }
+    migrations_to_apply
+}
+
+async fn get_migration_from_sql(
```

Comment:
When you take a flight from hometown, do you come FROM flight or come from hometown? Do you come by flight or come by hometown? Do you come from hometown by flight or come by hometown from flight?